File: python/tidegraph/settle.py
"""SettleGraph direct region-major execution and explicit TimedDAG embedding."""
from collections import defaultdict
from dataclasses import dataclass, replace
import torch
from .blocks import canonicalize, deliver, evaluate_block
from .graph import Edge, Graph, Node, Region
from .ops import Model
from .records import Atom, Continuation, External, Result
from .validation import validate_window


@dataclass(frozen=True)
class SettleGraph:
    graph: Graph
    ranks: tuple[int, ...]  # One strictly positive rank per region.

    def __post_init__(self):
        g = self.graph
        if (len(self.ranks) != len(g.regions) or len(set(self.ranks)) != len(self.ranks)
                or any(type(rank) is not int or rank <= 0 for rank in self.ranks)):
            raise ValueError("regions require distinct positive ranks")
        if not g.inputs or not g.outputs:
            raise ValueError("SettleGraph requires input and output boundaries")
        for e in g.edges:
            if e.delay != self.rank(e.target) - self.rank(e.source) or e.delay <= 0:
                raise ValueError("SettleGraph edges must follow region ranks with matching delays")

    def rank(self, node):
        return self.ranks[self.graph.nodes[node].region]

    @property
    def output_rank(self):
        return max(self.ranks) + 1

    @property
    def stride(self):
        return self.output_rank + 1

    def external(self, values, start_position=0, encoded=False):
        if values.ndim != 3:
            raise ValueError("values require [batch, sequence, width]")
        return [External(b, p, start_position + t,
                         self.stride * (start_position + t) + (0 if encoded else self.rank(v)), values[b, t])
                for b in range(len(values)) for p, v in enumerate((0,) if encoded else self.graph.inputs)
                for t in range(values.shape[1])]

    def embed(self, model):
        g = self.graph
        n, r = len(g.nodes), len(g.regions)
        edges = (g.edges + tuple(Edge(n, v, self.rank(v)) for v in g.inputs)
                 + tuple(Edge(v, n + 1, self.output_rank - self.rank(v)) for v in g.outputs))
        encoded = Graph(g.nodes + (Node(r, identity=True), Node(r + 1, identity=True)), edges,
                        g.regions + (Region(1), Region(1)), (n,), (n + 1,))
        em = Model(encoded, model.width, dtype=model.nodes[0].bias.dtype)
        em.nodes = torch.nn.ModuleList(list(model.nodes) + list(em.nodes[-2:]))
        def one():
            return torch.nn.Parameter(model.nodes[0].bias.new_ones(()), requires_grad=False)
        em.input_scale = torch.nn.ParameterList([one()])
        em.agg_scale = torch.nn.ParameterList(list(model.agg_scale) + list(model.input_scale) + list(model.output_scale))
        em.edge_scale = torch.nn.ParameterList(list(model.edge_scale) + [one() for _ in (*g.inputs, *g.outputs)])
        em.output_scale = torch.nn.ParameterList([one()])
        return encoded, em

    def embed_initial(self, q, encoded):
        if q.cut or q.pending or q.ledger or q.history:
            raise ValueError("embed_initial requires an initial cut; retain encoded continuation for subsequent windows")
        return replace(q.fork(), identity=encoded.identity)

    def project(self, result):
        if result.continuation.cut % self.stride or result.continuation.pending:
            raise ValueError("SettleGraph projection requires a complete position cut")
        g = self.graph
        n, r, e = len(g.nodes), len(g.regions), len(g.edges)
        def atom(a):
            if a.source < e:
                return a
            port = a.source - e
            if not 0 <= port < len(g.inputs):
                raise ValueError("invalid body input adapter edge")
            return Atom(a.batch, a.node, a.time, 0, port, a.position // self.stride, a.value)
        events = []
        for event in result.trace:
            if event["node"] < n:
                events.append(dict(event, fiber=[atom(a) for a in event["fiber"]]))
        q = result.continuation
        ledger = {(b, p): (position, time + self.rank(v)) for (b, _), (position, time) in q.ledger.items()
                  for p, v in enumerate(g.inputs)}
        projected = Continuation(g.identity, q.batch_size, q.cut, {k: s for k, s in q.states.items() if k[1] < n},
                                 {k: h for k, h in q.history.items() if k[1] < r},
                                 [a for a in q.pending if a.source < e], ledger)
        return Result(projected, events, result.outputs, [a for a in result.messages if a.source < e], result.stats)


def run(spec, model, q, values, *, mode="hard", zeta=1.0, prefill=True):
    """Independent SettleGraph schedule: each region settles the whole sequence."""
    if q.cut % spec.stride or q.pending or values.shape[0] != q.batch_size:
        raise ValueError("SettleGraph continuation must be a complete position boundary")
    start = q.cut // spec.stride
    stop = q.cut + values.shape[1] * spec.stride
    graph = spec.graph
    atoms, ledger = validate_window(graph, model, q, spec.external(values, start), stop, stop)
    q = q.fork(); q.ledger = ledger
    fibers = defaultdict(list)
    for a in atoms:
        fibers[a.batch, a.node, a.time].append(a)
    events, messages, raw_outputs = [], [], []
    stats = {"region_blocks": 0, "state_blocks": 0, "state_steps": 0, "full_blocks": 0}
    for region in sorted(range(len(graph.regions)), key=lambda r: spec.ranks[r]):
        members = {v for v, n in enumerate(graph.nodes) if n.region == region}
        frames = [(b, region, spec.stride * t + spec.ranks[region], members)
                  for t in range(start, start + values.shape[1]) for b in range(q.batch_size)]
        if not frames:
            continue
        block, counters = evaluate_block(graph, model, q, frames, fibers, mode=mode, zeta=zeta, prefill=prefill)
        deliver(graph, model, block, fibers, messages, raw_outputs)
        events.extend(block); stats["region_blocks"] += 1
        for key, value in counters.items():
            stats[key] = stats.get(key, 0) + value
    groups = defaultdict(list)
    for b, time, port, value in raw_outputs:
        position = (time - spec.rank(graph.outputs[port])) // spec.stride
        groups[b, position].append((port, value))
    outputs = []
    for (b, position), pairs in groups.items():
        values_out = [v for _, v in sorted(pairs)]
        value = values_out[0]
        for v in values_out[1:]:
            value = value + v
        outputs.append((b, position * spec.stride + spec.output_rank, 0, value))
    q.cut = stop
    result = canonicalize(graph, Result(q, events, outputs, messages, stats))
    result.outputs.sort(key=lambda o: (o[1], o[0]))
    return result

File: cpp/src/settle.cpp
#include "tide/settle.h"
#include "tide/kernel.h"
#include "tide/ops.h"
#include <algorithm>
#include <limits>
#include <set>
#include <stdexcept>

namespace tide {
SettleGraph::SettleGraph(Graph body, std::vector<Index> ranks)
    : body_(std::move(body)), ranks_(std::move(ranks)) {
  body_.compile();
  if (ranks_.size() != body_.regions.size() ||
      std::set<Index>(ranks_.begin(), ranks_.end()).size() != ranks_.size() ||
      std::any_of(ranks_.begin(), ranks_.end(), [](Index r) {
        return r <= 0 || r > std::numeric_limits<Index>::max() - 2;
      })) throw std::invalid_argument("SettleGraph requires distinct positive representable region ranks");
  if (body_.inputs.empty() || body_.outputs.empty())
    throw std::invalid_argument("SettleGraph requires input and output boundaries");
  stride_ = *std::max_element(ranks_.begin(), ranks_.end()) + 2;
  for (const auto& edge : body_.edges)
    if (edge.delay != rank(edge.target) - rank(edge.source))
      throw std::invalid_argument("SettleGraph edges must follow region ranks with matching delays");
  encoded_ = body_;
  const Index n = body_.nodes.size(), r = body_.regions.size();
  encoded_.nodes.push_back(Node{r, false, true});
  encoded_.nodes.push_back(Node{r + 1, false, true});
  encoded_.regions.push_back(Region{1});
  encoded_.regions.push_back(Region{1});
  auto& layout = *encoded_.layout;
  auto& domain = *encoded_.source_domain;
  for (Index p = 0; p < static_cast<Index>(body_.inputs.size()); ++p) {
    const auto node = body_.inputs[p];
    const Index edge = encoded_.edges.size();
    encoded_.edges.push_back({n, node, rank(node)});
    layout.edge_source.push_back(p);
    layout.edge_target.push_back(body_.layout->input[p]);
    domain.edge_target.push_back(body_.source_domain->input[p]);
    encoded_.origins.push_back({edge, p, stride_});
  }
  for (Index p = 0; p < static_cast<Index>(body_.outputs.size()); ++p) {
    const auto node = body_.outputs[p];
    encoded_.edges.push_back({node, n + 1, output_rank() - rank(node)});
    layout.edge_source.push_back(body_.layout->output[p]);
    layout.edge_target.push_back(p);
    domain.edge_target.push_back(p);
  }
  encoded_.inputs = {n}; encoded_.outputs = {n + 1};
  layout.input = {0}; layout.output = {0}; domain.input = {0};
  encoded_.compile();
  encoded_.topological_order();
}

Index SettleGraph::rank(Index node) const { return ranks_.at(body_.nodes.at(node).region); }

Model SettleGraph::embed_model(const Model& body) const {
  Model result = body;
  configure_model(body_, result); validate_model(body_, result);
  const auto options = body.nodes[0].bias.options().requires_grad(false);
  for (int i = 0; i < 2; ++i) {
    result.nodes.push_back({at::zeros({body.width()}, options), at::zeros({body.width(), body.width()}, options),
                            at::zeros({body.width()}, options), at::zeros({body.width()}, options)});
    result.regions.push_back({});
  }
  auto one = [&] { return at::ones({}, options); };
  result.input_scale = {one()}; result.output_scale = {one()};
  result.agg_scale.insert(result.agg_scale.end(), body.input_scale.begin(), body.input_scale.end());
  result.agg_scale.insert(result.agg_scale.end(), body.output_scale.begin(), body.output_scale.end());
  for (size_t i = 0; i < body_.inputs.size() + body_.outputs.size(); ++i) result.edge_scale.push_back(one());
  configure_model(encoded_, result); validate_model(encoded_, result);
  return result;
}

Continuation SettleGraph::embed_initial(const Continuation& initial) const {
  if (initial.identity != body_.identity || initial.batch_size < 1 || initial.cut != 0 ||
      !initial.pending.empty() || !initial.ledger.empty())
    throw std::invalid_argument("SettleGraph embed_initial requires body identity and an initial cut");
  auto owners = [&](const auto& values, Index count) {
    for (const auto& item : values) {
      const auto [b, v] = item.first;
      if (b < 0 || b >= initial.batch_size || v < 0 || v >= count)
        throw std::invalid_argument("invalid body initial owner in SettleGraph embedding");
    }
  };
  owners(initial.states, body_.nodes.size()); owners(initial.history, body_.regions.size());
  auto result = initial; result.identity = encoded_.identity;
  return result;
}

std::vector<External> SettleGraph::external(const Tensor& values, Index start, bool encoded) const {
  if (!values.defined() || values.dim() != 3 || values.size(0) < 1 || values.size(2) < 1)
    throw std::invalid_argument("SettleGraph values require [batch, sequence, width]");
  // Bound the complete stop as well as every event, without intermediate overflow.
  if (start < 0 || start > std::numeric_limits<Index>::max() / stride_ ||
      values.size(1) > std::numeric_limits<Index>::max() / stride_ - start)
    throw std::invalid_argument("SettleGraph position clock overflow");
  std::vector<External> result;
  const Index ports = encoded ? 1 : body_.inputs.size();
  for (Index b = 0; b < values.size(0); ++b)
    for (Index p = 0; p < ports; ++p)
      for (Index t = 0; t < values.size(1); ++t) {
        const auto position = start + t;
        result.push_back({b, p, position, stride_ * position + (encoded ? 0 : rank(body_.inputs[p])), values[b][t]});
      }
  return result;
}

SettleExecutor::SettleExecutor(SettleGraph spec, Model model, Options options, std::string algorithm)
    : spec_(std::move(spec)) {
  if (algorithm == "frontier")
    frontier_ = std::make_unique<Frontier>(spec_.encoded_graph(), spec_.embed_model(model), options);
  else if (algorithm == "streaming")
    streaming_ = std::make_unique<Streaming>(spec_.encoded_graph(), spec_.embed_model(model), options);
  else throw std::invalid_argument("SettleExecutor algorithm must be frontier or streaming");
}

Result SettleExecutor::run(const Continuation& initial, const Tensor& values) {
  if (initial.cut < 0 || initial.cut % spec_.stride() || !initial.pending.empty())
    throw std::invalid_argument("SettleGraph run requires a complete position boundary");
  const auto start = initial.cut / spec_.stride();
  auto inputs = spec_.external(values, start);
  if (initial.batch_size != values.size(0)) throw std::invalid_argument("SettleGraph batch mismatch");
  const auto stop = (start + values.size(1)) * spec_.stride();
  return frontier_ ? frontier_->run(initial, inputs, stop, stop) : streaming_->run(initial, inputs, stop, stop);
}
}  // namespace tide
